paddle/cinn/optim/elinimate_common_factor_of_local_index.cc

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2024 CINN Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/cinn/optim/elinimate_common_factor_of_local_index.h"
+
+#include <unordered_map>
+
+#include "paddle/cinn/common/cas.h"
+#include "paddle/cinn/ir/ir.h"
+#include "paddle/cinn/ir/ir_mutator.h"
+#include "paddle/cinn/ir/ir_printer.h"
+#include "paddle/cinn/ir/utils/ir_copy.h"
+#include "paddle/cinn/utils/string.h"
+
+namespace cinn {
+namespace optim {
+namespace {
+
+class GatherLocalIndexVisitor : public ir::IRMutator<> {
+ public:
+  void operator()(ir::Expr* expr) { ir::IRMutator<>::Visit(expr, expr); }
+
+  const std::unordered_map<std::string, std::vector<std::vector<ir::Expr>>>&
+  local_var_to_indexes() {
+    return local_var_to_indexes_;
+  }
+
+ private:
+  void Visit(const ir::Store* op, Expr* expr) override {
+    auto store = expr->As<ir::Store>();
+
+    ir::IRMutator<>::Visit(op, expr);
+    if (!store->tensor.as_tensor_ref()->buffer.defined()) {
+      return;
+    }
+
+    if (store->tensor.as_tensor_ref()->buffer->memory_type ==
+        ir::MemoryType::GPULocal) {
+      local_var_to_indexes_[store->tensor.as_tensor_ref()->buffer->name]
+          .push_back(store->indices);
+    }
+  }
+
+  void Visit(const ir::Load* op, Expr* expr) override {
+    auto load = expr->As<ir::Load>();
+
+    if (load->is_addr_scalar()) {
+      return;
+    }
+    if (!load->tensor.as_tensor_ref()->buffer.defined()) {
+      return;
+    }
+
+    if (load->tensor.as_tensor_ref()->buffer->memory_type ==
+        ir::MemoryType::GPULocal) {
+      local_var_to_indexes_[load->tensor.as_tensor_ref()->buffer->name]
+          .push_back(load->indices);
+    }
+    ir::IRMutator<>::Visit(op, expr);
+  }
+
+  std::unordered_map<std::string, std::vector<std::vector<ir::Expr>>>
+      local_var_to_indexes_;
+};
+
+template <typename DoEachT>
+void VisitEachRowExpr(const std::vector<std::vector<ir::Expr>>& indexes,
+                      std::size_t var_idx,
+                      DoEachT&& DoEach) {
+  for (std::size_t i = 0; i < indexes.size(); ++i) {
+    DoEach(indexes[i][var_idx]);
+  }
+}
+
+int ExtractNumberFromExpr(const ir::Expr& expr) {
```

Comment:
It seems some size, index can be int64, is `int` enough for cases here?

Reply:
This function only handles the factor of ir::Expr, so we will not meet the int64 problem in most cases. Besides, considering int64 may affect the kernel performance, use `int` as the function result.